### ISSUE TITLE
Creating a LeafletJs Map WIdget

### ISSUE THREAD

**Issue by stefanomonteiro** [elementor/elementor#8939]
I am trying to create a leaflet map widget. Managed to make it work on the frontend, however facing some issues on the editor preview. Before getting into my issue I would like to discuss the best way to create the widget. To create the map we need to write some JavaScript ([see a quick tutorial here](https://leafletjs.com/examples/quick-start/)). That is the challenge of creating the elementor widget. I thought of `echo` it into `<script>` tags on the `function render()`, it works but it requires to load the `leaflet.js` in the header (not ideal). Is there a better solution?

Moving forward to my issue:

When I make any change on the panel the map (widget) disappears. I have tried different ways to write the `function _content_template()`, here is how I have it currently (no error shown in the browser console).

```
 protected function _content_template()
    {
        ?>

        <div id="map" class="picante__map"></div>
        <script src="https://unpkg.com/leaflet@1.5.1/dist/leaflet.js" integrity="sha512-GffPMF3RvMeYyc1LWMHtK8EbPv0iNZ8/oTtHPx9/cc2ILxQ+u905qIwdpULaqDkyBKgOaB57QTMg7ztg8Jm2Og==" crossorigin=""></script>
        <# 
    var mymap=L.map('map').setView([{{{settings.lat}}}, {{{settings.lon}}}], {{{settings.zoom}}}); 
    
    L.tileLayer('https://api.tiles.mapbox.com/v4/{id}/{z}/{x}/{y}.png?access_token={{{settings.api_key}}}', { attribution: '' , maxZoom: 18, id: 'mapbox.{{{settings.map_style}}}', accessToken: 'your.mapbox.access.token' }).addTo(mymap); 
    
    var newIcon = L.icon({
                iconUrl: {{{settings.pin.url}}},
                iconSize: [{{{setting.pin_size}}}, {{{setting.pin_size}}}]
            })

    if ({{{settings.pins}}}) {
        var counter = 0;
        _.each({{{settings.pins}}}, function(pin) {
            counter++;

            L.marker([{{{pin.pin_lat}}}, {{{pin.pin_lon}}}], {
                icon: newIcon
            }).addTo(mymap).bindPopup('<h3>{{{pin.pin_title}}}</h3>{{{pin.pin_content}}}');
        });
    };


    if ('yes' === {{{settings.get_coord}}}) {

        var popup = L.popup();

        function onMapClick(e) {
            popup
                .setLatLng(e.latlng)
                .setContent('Latitude: ' + e.latlng.lat + ' <br> Longitude: ' + e.latlng.lng)
                .openOn(mymap);
        }
        mymap.on('click', onMapClick);

    }
    #>


<?php
    }

```

Another thing that doesn't look ideal is to have to load `leaflet.js` again as it is already being enqueued in the header.

To summarize my questions:

1 - What is the best way to write this widget as we need to add the `controls` input into javascript;

2 - What in the above code is making the widget disappear from the editor preview;

Thanks

**Comment by bainternet** [comment]
Please see our tutorial on adding JS handlers to your widgets: https://developers.elementor.com/building-a-simple-custom-widget-with-javascript/

**Comment by stefanomonteiro** [comment]
Thanks @bainternet.

After some more digging, I believe I am almost finished. Only need to understand why a repeater object is returning different on the frontend from the backend.

On the backend, it returns:
```
{
 length: 4
model: ƒ (t,n)
models: (4) [N.d, N.d, N.d, N.d]
_byId: {c26: N.d, c27: N.d, c28: N.d, c29: N.d}
}
```

On the frontend, it returns the array of repeated items directly:
```
{
 (4) [{…}, {…}, {…}, {…}]
}
```

Here is my code:
```
class StemonMap extends elementorModules.frontend.handlers.Base {
  getDefaultSettings() {
    return {
      controls: {
         ...
        pins: this.getElementSettings("pins"),
        ...
      }
    };
  }

  getDefaultElements() {
    const controls = this.getSettings("controls");
    return {
      ...
      pins: this.$element.find(controls.pins),
      ...
    };
  }

  bindEvents() {
    const set = {
      ...
      pins: this.elements.pins.selector, // Frontend
      pins: this.elements.pins.selector.models, // Backend
      ...
    };
          ...
    set.pins.forEach(pin => {
      console.log(pin.pin_title); //  Frontend
      console.log(pin.attributes.pin_title); // Backend    });
  }
}

jQuery(window).on("elementor/frontend/init", () => {
  const addHandler = $element => {
    elementorFrontend.elementsHandler.addHandler(StemonMap, {
      $element
    });
  };

  elementorFrontend.hooks.addAction(
    "frontend/element_ready/leaflet-map.default",
    addHandler
  );
});

```

Any suggestion on how to go around this?

Thanks again
